core/src/main/java/io/substrait/relation/RelCopyOnWriteVisitor.java

```diff
@@ -102,6 +102,20 @@ public Optional<Rel> visit(Join join) throws RuntimeException {
             .build());
   }
 
+  @Override
+  public Optional<Rel> visit(Set set) throws RuntimeException {
+    var inputs =
+        set.getInputs().stream()
+            .map(inputRel -> inputRel.accept(this))
+            .flatMap(Optional::stream)
+            .toList();
```

Comment:
There are two use cases for the RelCopyOnWriteVisitor:

1. To traverse the tree to detect something. For example, count how many occurrences of COUNT DISTINCT there are across the entire tree.
2. To rewrite parts of the subtree. For example, to replace any occurrence of COUNT DISTINCT with a call to APPROX_COUNT_DISTINCT instead.

For the default read-only traversal case of (1), each visitor method returns Optional.empty() as an indication to the caller that nothing has changed. That way the caller knows not to create a new instance of the Rel. Only for the rewrite case of (2) will a new Set relation be created. 

In support of this pattern, there's a `RelCopyOnWriteVisitor.transformList` static function to help. For an example usage, see [here](https://github.com/substrait-io/substrait-java/blob/5872ceaaef67b8cb774017a50e0f2920fe0a56f3/isthmus/src/test/java/io/substrait/isthmus/RelCopyOnWriteVisitorTest.java#L228-L234). You'll want to do something similar to only create the new Set if one of the inputs changed.

Reply:
Thanks for the explanation.
I have fixed the function to return Optional.empty() only if all of the inputs didn't change.
Note that I preferred doing two passes over the input list since doing it in one pass in functional programming using streams would have introduced a side effect which I would rather avoid as it's not best practice.
I expect the list of inputs to be reasonably small so a two pass won't be an issue.